mini-projects/01-device-driver/laptop-interface/src/main.rs

```diff
@@ -1,3 +1,233 @@
-fn main() {
-    println!("Hello, World!");
+mod cffi;
+mod gui;
+mod serial;
+
+use cffi::*;
+use gui::*;
+use serial::*;
+use std::io::{self};
+use std::sync::mpsc::{channel, Receiver};
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::Duration;
+
+fn main() -> Result<(), eframe::Error> {
+    println!("Do you want to run the onboarding project? (y/n):");
+
+    // Get the user's input
+    let mut input = String::new();
+    io::stdin()
+        .read_line(&mut input)
+        .expect("Failed to read input");
+    let input = input.trim().to_lowercase();
+
+    if input == "y" {
+        // Run the onboarding project: connect to two controllers
+        println!("Starting the onboarding project...");
+        run_onboarding_project();
+    } else {
+        // Run the GUI for a single controller
+        println!("Starting the GUI with a single controller...");
+
+        let circle_state = Arc::new(Mutex::new(CircleState::new()));
+
+        // Get available ports
+        let port_name = unsafe {
+            match get_ports() {
+                Ok(mut ports) if !ports.is_empty() => ports.remove(0),
+                Ok(_) => {
+                    eprintln!("No USB controllers found for GUI.");
+                    return Ok(());
+                }
+                Err(err) => {
+                    eprintln!("Failed to list ports: {}", err);
+                    return Ok(());
+                }
+            }
+        };
+
+        // Start the serial thread for the single controller
+        let state_clone = circle_state.clone();
+        thread::spawn(move || unsafe {
+            match init_serial(&port_name) {
+                Ok(port_handle) => {
+                    blocking_write(port_handle, b"start controller\n").ok();
+                    loop {
+                        if let Some(controller_state) = read_controller_state(port_handle) {
+                            let mut state = state_clone.lock().unwrap();
+                            state.update(controller_state.buttons, controller_state.adc_value);
+                        }
+                    }
+                }
+                Err(err) => {
+                    eprintln!("Failed to initialize serial port: {}", err);
+                }
+            }
+        });
```

Comment:
I think your code could have benefitted from having a more explicit "controller manager" struct, it would move a lot of this USB interface logic out of this main function, cleaning it up, and making it more explicit where to debug controller issues